src/entity_embedding_utils.jl

```diff
@@ -0,0 +1,148 @@
+"""
+
+EMBEDDING_ENABLED_MODELS = [
+    NeuralNetworkClassifier,
+    NeuralNetworkBinaryClassifier,
+    NeuralNetworkRegressor,
+    MultitargetNeuralNetworkRegressor,
+]
+
+EMBEDDING_ENABLED_MODELS_UNION = Union{EMBEDDING_ENABLED_MODELS...}
+
+
+# A function to check if a model is among those in EMBEDDING_ENABLED_MODELS
+function is_embedding_enabled_type(model_type)
+    return any(model_type <: T for T in EMBEDDING_ENABLED_MODELS)
+end
+
+# function to set default new embedding dimension 
+function set_default_new_embedding_dim(numlevels)
+    # Either min_ratio or max_ratio of numlevels depending on >= threshold or < threshold
+    min_ratio, max_ratio = 0.2, 0.5
+    threshold = 20
+    return ceil(Int, ((numlevels >= threshold) ? min_ratio : max_ratio) * numlevels)
+end
+
+MISMATCH_INDS(wrong_feats) =
+    "Features $(join(wrong_feats, ", ")) were specified in embedding_dims hyperparameter but were not recognized as categorical variables because their scitypes are not `Multiclass` or `OrderedFactor`."
+function check_mismatch_in_cat_feats(featnames, cat_inds, specified_featinds)
+    wrong_feats = [featnames[i] for i in specified_featinds if !(i in cat_inds)]
+    length(wrong_feats) > 0 && throw(ArgumentError(MISMATCH_INDS(wrong_feats)))
+end
+
+# function to set new embedding dimensions
+function set_new_embedding_dims(featnames, cat_inds, num_levels, embedding_dims)
+    specified_featnames = keys(embedding_dims)
+    specified_featinds =
+        [i for i in 1:length(featnames) if featnames[i] in specified_featnames]
+    check_mismatch_in_cat_feats(featnames, cat_inds, specified_featinds)
+    catind2numlevels = Dict(zip(cat_inds, num_levels))
+    # for each value of embedding dim if float then multiply it by the number of levels
+    for featname in specified_featnames
+        if embedding_dims[featname] isa AbstractFloat
+            embedding_dims[featname] = ceil(
+                Int,
+                embedding_dims[featname] *
+                catind2numlevels[findfirst(x -> x == featname, featnames)],
+            )
+        end
+    end
+    newdims = [
+        (cat_ind in specified_featinds) ? embedding_dims[featnames[cat_ind]] :
+        set_default_new_embedding_dim(num_levels[i]) for
+        (i, cat_ind) in enumerate(cat_inds)
+    ]
+    return newdims
+end
+
+
+"""
+**Private Method**
+
+Returns the indices of the categorical columns in the table `X`.
+"""
+function get_cat_inds(X)
+    # if input is a matrix; conclude no categorical columns
+    Tables.istable(X) || return Int[]
+    types = schema(X).scitypes
+    cat_inds = findall(x -> x <: Finite, types)
+    return cat_inds
+end
+
+"""
+**Private Method**
+
+Returns the number of levels in each categorical column in the table `X`.
+"""
+function get_num_levels(X, cat_inds)
+    num_levels = []
+    for i in cat_inds
+        num_levels = push!(num_levels, length(levels(Tables.getcolumn(X, i))))
+    end
```

Comment:
Have you taken steps to ensure `X` implements the Tables `AbstractColumns` interface? Not every table does, and you cannot reliably call `Tables.column` on an arbitrary table:

```julia
help?> Tables.getcolumn
  Tables.getcolumn(::AbstractColumns, nm::Symbol) => Indexable collection with known length
  Tables.getcolumn(::AbstractColumns, i::Int) => Indexable collection with known length
  Tables.getcolumn(::AbstractColumns, T, i::Int, nm::Symbol) => Indexable collection with known length
  
  Tables.getcolumn(::AbstractRow, nm::Symbol) => Column value
  Tables.getcolumn(::AbstractRow, i::Int) => Column value
  Tables.getcolumn(::AbstractRow, T, i::Int, nm::Symbol) => Column value

  Retrieve an entire column (from AbstractColumns) or single row column value
  (from an AbstractRow) by column name (nm), index (i), or if desired, by
  column element type (T), index (i), and name (nm). When called on a
  AbstractColumns interface object, the returned object should be a 1-based
  indexable collection with known length. When called on a AbstractRow
  interface object, it returns the single column value. The methods taking a
  single Symbol or Int are both required for the AbstractColumns and
  AbstractRow interfaces; the third method is optional if type stability is
  possible. The default definition of Tables.getcolumn(x, i::Int) is
  getfield(x, i). The default definition of Tables.getcolumn(x, nm::Symbol) is
  getproperty(x, nm).
```

Reply:
OK. I accordingly handled all instances of `Tables.getcolumn(X, i)` by using `Tables.columns(X)` as in 46577f714b911f32b6688df0849be84d1b6eaedb.